File: rails/lib/eng_lib.rb
class EngLib
  @conn = Faraday.new(
    url: "https://www.library.eng.chula.ac.th",
    headers: {'Content-Type' => 'application/x-www-form-urlencoded'},
    ) do |conn|
      conn.request  :url_encoded
      conn.response :json, content_type: /\bjson$/
  end
  def self.user(id)
    resp = @conn.post("/api/user/verify-student", {"studentID": id})
    body = resp.body.with_indifferent_access
    {
      :id         => body["studentId"],
      :nameEN     => body["fName"],
      :surnameEN  => body["lName"],
      :email      => body["email"],
      :department => body["department"],
      :tel        => body["tel"]
    } if body["message"] == "valid"
  end
end

File: nestjs/src/lib/englib.ts
import { NotFoundException } from '@nestjs/common';
import { map, pluck, tap, retry } from 'rxjs/operators';
import axios from 'axios';
import { defer } from 'rxjs';
import { Person } from '../entity/person';

export function englib(studentId: string) {
  return defer(() =>
    axios.request({
      method: 'POST',
      url: '/api/user/verify-student',
      baseURL: 'https://www.library.eng.chula.ac.th',
      data: { studentID: studentId },
    }),
  )
    .pipe(
      retry(3),
      pluck('data'),
      tap((raw) => {
        if (raw.message == 'invalid')
          throw new NotFoundException('user not exist in cu engineer api');
      }),
      map<any, Partial<Person>>((raw) => ({
        student_id: raw.studentId,
        name_en: raw.fName,
        surname_en: raw.lName,
        email: raw.email,
        department: raw.department,
        phone: raw.tel,
      })),
    )
    .toPromise();
}
